File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.py
class Solution:
    def helper(self,nums,target):
        cnt=0
        st=0
        e=0
        preSum=0
        while e<len(nums):
            preSum=preSum+nums[e]
            if preSum<=target:
                cnt=cnt+(e-st+1)
                e=e+1
            else:
                while preSum>target and st<=e:
                    preSum=preSum-nums[st]
                    st=st+1
                if preSum<=target:
                    cnt=cnt+(e-st+1)
                e=e+1
        return cnt

    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        ans1=self.helper(nums, goal)
        ans2=self.helper(nums,goal-1)
        return ans1-ans2

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    public int helper(int []nums,int k){
        if(k<0){
            return 0;
        }
        int length=nums.length;
        int left=0;
        int right=0;
        int count=0;
        int sum=0;
        while(right<length){
            sum+=nums[right];
            while(sum>k){
                sum-=nums[left];
                left=left+1;
            }
            count+=(right-left+1);
            right++;
        }
        return count;
    }
    public int numSubarraysWithSum(int[] nums, int goal) {
        return helper(nums,goal)-helper(nums,goal-1);
    }
}